### DISCUSSION TITLE
Default solver (pypardiso) not working on AMD cpus

### DISCUSSION THREAD

**Discussion by weishiyu1** [PMEAL/OpenPNM#2272] [Q&A] [answered]
I have encountered such a problem (`AttributeError: module 'openpnm' has no attribute 'models'`) when I import openpnm. How can I solve it?

**Comment by ma-sadeghi** [comment, marked as answer]
It might just be a corrupt installation. Try re-installing `openpnm`.

**Reply by weishiyu1** [reply]
I reinstalled it again with pip way, however, it still doesn’t seem to work

**Reply by jgostick** [reply]
This error really shouldn't be happening.  The models module has been part of openpnm for several years so it's not coming from the new V3 changes we've made.  What about doing ``pip uninstall openpnm`` first?  Also, can you check that your doing pip install from the conda prompt rather than the windows prompt?   Also, make sure your installing into the same environment that your running it from.  
You can check this by typing ``python`` at the command prompt to enter python then to ``import openpnm``.

**Reply by ma-sadeghi** [reply]
Here's how you can install `openpnm` in a new virtual env: Open Anaconda prompt and then run:

1. `conda create -n testenv python=3.8
2. `conda activate testenv`
3. `conda install openpnm`
4. python -c "import openpnm.models"

**Reply by weishiyu1** [reply]
I installed it again according to the above steps, the problem was solved, but there is a new problen and  permeability could not be calculated (Intel MKL FATAL ERROR: Cannot load mkl_intel_thread.dll.)

**Reply by ma-sadeghi** [reply]
Could you please paste the entire error?

**Reply by weishiyu1** [reply]
> Could you please paste the entire error?

openpnm.network.GenericNetwork : berea
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
#     Properties                                    Valid Values
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1     pore.area                                     27004 / 27004
2     pore.clay_volume                              27004 / 27004
3     pore.coords                                   27004 / 27004
4     pore.radius                                   27004 / 27004
5     pore.shape_factor                             27004 / 27004
6     pore.volume                                   27004 / 27004
7     throat.area                                   75804 / 75804
8     throat.clay_volume                            75804 / 75804
9     throat.conduit_lengths.pore1                  75804 / 75804
10    throat.conduit_lengths.pore2                  75804 / 75804
11    throat.conduit_lengths.throat                 75804 / 75804
12    throat.conns                                  75804 / 75804
13    throat.length                                 75804 / 75804
14    throat.radius                                 75804 / 75804
15    throat.shape_factor                           75804 / 75804
16    throat.total_length                           75804 / 75804
17    throat.volume                                 75804 / 75804
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
#     Labels                                        Assigned Locations
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1     pore.all                                      27004     
2     pore.geo_01                                   27004     
3     pore.inlets                                   900       
4     pore.outlets                                  900       
5     throat.all                                    75804     
6     throat.geo_01                                 75804     
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Number of pores before trimming:  27004
Number of pores after trimming:  26056
[7.7746e-25 1.2105e-24 2.0507e-24 ... 3.0959e-24 3.3230e-18 3.3230e-18]
Intel MKL FATAL ERROR: Cannot load mkl_intel_thread.dll.

Process finished with exit code 2

**Reply by weishiyu1** [reply]
Perhaps this is a problem associate with the processor? My computer is AMD ryzen processor, mkl_intel_thread.dll is a  package from intel?

**Reply by ma-sadeghi** [reply]
Oh, I see! Could be... The default solver is pypardiso which only works if you have MKL installed. But, I think that MKL should work fine on AMD processors, it's just more optimized on Intel ones. Just to make sure that's the case, when you run your algorithm, i.e., call `alg.run()`, do this instead:

```python
alg.set_solver(solver_family="scipy", solver_type="spsolve")
alg.run()
```
Btw, this api is going to slightly change in v3.

**Reply by weishiyu1** [reply]
> Oh, I see! Could be... The default solver is pypardiso which only works if you have MKL installed. But, I think that MKL should work fine on AMD processors, it's just more optimized on Intel ones. Just to make sure that's the case, when you run your algorithm, i.e., call `alg.run()`, do this instead:
> 
> ```python
> alg.set_solver(solver_family="scipy", solver_type="spsolve")
> alg.run()
> ```
> 
> Btw, this api is going to slightly change in v3.

Thank you so much, this problem has been solved.

**Reply by ma-sadeghi** [reply]
@weishiyu1 We're trying to investigate this issue more and if needed, add a fallback mechanism for AMD machines. Could you please paste the "entire" error traceback (the one that inclues which lines led to the error, since the one you've pasted here doesn't include those information) when using the default solver?

**Reply by weishiyu1** [reply]
> @weishiyu1我们正在尝试更多地调查这个问题，如果需要，为 AMD 机器添加回退机制。您能否在使用默认求解器时粘贴“整个”错误回溯（包括哪些行导致错误，因为您在此处粘贴的行不包含这些信息）？

```python
Traceback (most recent call last):
  File "E:/007_Code/OpenPNM/permeability_calculation.py", line 30, in <module>
    flow.run()
  File "C:\Users\Yoorken\anaconda3\envs\test\lib\site-packages\openpnm\algorithms\ReactiveTransport.py", line 159, in run
    x = self._run_reactive(x0)
  File "C:\Users\Yoorken\anaconda3\envs\test\lib\site-packages\openpnm\algorithms\ReactiveTransport.py", line 361, in _run_reactive
    self[quantity] = x = self._solve(x0=x) * w + x * (1 - w)
  File "C:\Users\Yoorken\anaconda3\envs\test\lib\site-packages\openpnm\algorithms\GenericTransport.py", line 621, in _solve
    x = solver(A, b, atol=atol, rtol=rtol, max_it=max_it, x0=x0)
  File "C:\Users\Yoorken\anaconda3\envs\test\lib\site-packages\openpnm\algorithms\GenericTransport.py", line 680, in solver
    x = pypardiso.spsolve(A=A, b=b)
  File "C:\Users\Yoorken\anaconda3\envs\test\lib\site-packages\pypardiso\scipy_aliases.py", line 46, in spsolve
    solver.factorize(A)
  File "C:\Users\Yoorken\anaconda3\envs\test\lib\site-packages\pypardiso\pardiso_wrapper.py", line 150, in factorize
    self._call_pardiso(A, b)
  File "C:\Users\Yoorken\anaconda3\envs\test\lib\site-packages\pypardiso\pardiso_wrapper.py", line 263, in _call_pardiso
    self._mkl_pardiso(self.pt.ctypes.data_as(ctypes.POINTER(self._pt_type[0])),  # pt
OSError: [WinError -1066598274] Windows Error 0xc06d007e
```

**Reply by ma-sadeghi** [reply]
@weishiyu1 Thanks, this seems to be different error, right? You previously reported `Intel MKL FATAL ERROR: Cannot load mkl_intel_thread.dll.`. Is that correct or am I missing something?

**Reply by weishiyu1** [reply]
> @weishiyu1 Thanks, this seems to be different error, right? You previously reported `Intel MKL FATAL ERROR: Cannot load mkl_intel_thread.dll.`. Is that correct or am I missing something?

This is another running environment. You can ignore it. After configuring the environment and install openpnm package according to previous prompts, when it runs to the flow.run() , no other errors are displayed, just showed Intel MKL FATAL ERROR: Cannot load mkl_intel_thread.dll, and then it ends the operation.

**Reply by ma-sadeghi** [reply]
Hmm, does it say which line the error comes from?

**Reply by weishiyu1** [reply]
There is no prompt on which line the error is, only `Intel MKL FATAL ERROR: Cannot load mkl_intel_thread.dll` was displayed. It should be that the program has terminated, but according to my judgment, it might be come from `flow.run()`.

**Reply by ma-sadeghi** [reply]
Could you try to run only the following script and see what error (if any) you're getting?
```python
import pypardiso
import numpy as np
import scipy.sparse as sp
A = sp.rand(10, 10, density=0.5, format='csr')
b = np.random.rand(10)
x = pypardiso.spsolve(A, b)
```

**Reply by weishiyu1** [reply]
> Could you try to run only the following script and see what error (if any) you're getting?
> 
> ```python
> import pypardiso
> import numpy as np
> import scipy.sparse as sp
> A = sp.rand(10, 10, density=0.5, format='csr')
> b = np.random.rand(10)
> x = pypardiso.spsolve(A, b)
> ```

```python
Traceback (most recent call last):
  File "E:/007_Code/OpenPNM/permeability_calculation.py", line 6, in <module>
    x = pypardiso.spsolve(A, b)
  File "C:\Users\Yoorken\anaconda3\envs\testenv\lib\site-packages\pypardiso\scipy_aliases.py", line 46, in spsolve
    solver.factorize(A)
  File "C:\Users\Yoorken\anaconda3\envs\testenv\lib\site-packages\pypardiso\pardiso_wrapper.py", line 150, in factorize
    self._call_pardiso(A, b)
  File "C:\Users\Yoorken\anaconda3\envs\testenv\lib\site-packages\pypardiso\pardiso_wrapper.py", line 263, in _call_pardiso
    self._mkl_pardiso(self.pt.ctypes.data_as(ctypes.POINTER(self._pt_type[0])),  # pt
OSError: [WinError -1066598274] Windows Error 0xc06d007e
```
